### ISSUE TITLE
should all popup use cases have aria-haspopup on the trigger element ?

### ISSUE THREAD

**Issue by gaurav5430** [w3c/aria-practices#1926]
In our application, we have a popup which is triggered on click on of a button. The content in this popup is divided into 2 sections:
- the first section has some static + interactive elements
- the second section is a list of links

I would like to understand
-  if I should be using `aria-haspopup` on the trigger button ?
- Is it supposed to be used for all popup use cases?
- If not, then how do I correctly expose accessibility semantics in this case?
- If yes, what role should i use in `aria-haspopup` ?

This is how my example looks like:
<img width="168" alt="example of popup" src="https://user-images.githubusercontent.com/12510548/120119318-aca99100-c1b4-11eb-8247-490b34ab399b.png">

**Comment by JAWS-test** [comment]
Actually, for a pop-up on the triggering button aria-haspopup=dialog should be used. But there are two problems with this:
* the role in the aria-haspopup is often [not output](https://scottaohara.github.io/tests/aria-haspopup/).
* aria-haspopup at a button turns it into a menu button according to the [specification](https://github.com/w3c/core-aam/issues/51) - and this is output by the screen readers, even if it is not a menu button at all.

Therefore there are four possibilities:
* use aria-haspopup=dialog at the button, because this is correct according to the ARIA specification
* don't use aria-haspopup, because sighted users don't see what the button is doing either
* use aria-haspopup=dialog at the link to avoid the menu button problem
* don't use aria-haspopup, but give the button a description (title, aria-describedby or aria-description) that refers to the pop-up

See: https://github.com/w3c/aria/issues/1024

**Comment by gaurav5430** [comment]
the popup that I have, in the attached image, does not seem like a dialog. In fact, it more resembles a menu. What would be the criteria to treat it as a `dialog`?

**Comment by JAWS-test** [comment]
Your example shows a pop-up. A menu would be a list marked with role=menu and containing only items that are
* are marked with role=menuitem
* are only accessible with the arrow keys

Other elements must not be included in a menu.

**Comment by gaurav5430** [comment]
ok, but my use case does not seem like a dialog as well, in the typical sense atleast

also, hypothetically if I remove the first section of my popup, would we now consider it to be a menu?

**Comment by JAWS-test** [comment]
> if I remove the first section of my popup, would we now consider it to be a menu?

only, if you use role=menu and menuitem

**Comment by gaurav5430** [comment]
apologies if it was not clear already

I can still not understand why the `dialog` role would be more suited for this and not the `menu` role ?
I would not add `role="menuitem"` on any of the non interactive elements, but even then I am still confused for the other interactive elements in the first section.
 
Here is an example from twitter implementation, which has `role="menu"` and `"menuitem"` , do you think this should also be a `dialog` ?

<img width="383" alt="Screenshot 2021-05-31 at 1 05 57 AM" src="https://user-images.githubusercontent.com/12510548/120226974-f0a89e80-c265-11eb-9a8f-17db954ec00a.png">
<img width="330" alt="Screenshot 2021-05-31 at 1 06 29 AM" src="https://user-images.githubusercontent.com/12510548/120226979-f30af880-c265-11eb-8905-3082b5d604e3.png">

**Comment by JAWS-test** [comment]
> do you think this should also be a dialog?

* No, aria-haspopup=menu would be correct if it is a menu only, i.e. all content is in the menu and accessible via arrow key operation. 
* However, if e.g. the heading "How shows you top Tweets first" is not automatically output when opening the dialog or with the menuitems, then aria-haspopup=dialog would be correct, because then it is not a pure menu, but a dialog that contains a menu.
* If the items are marked as a menuitem, but cannot be operated like a menu, it would be wrong anyway.

**Comment by gaurav5430** [comment]
Thanks @mcking65 
just to sum up my understanding of what you suggested:

You are suggesting to use neither the `menu` role or the `dialog` role for my use case. 
If that is the case, would using `aria-expanded` and focusing on the popup when opened be enough for a screenreader user?

You also seem to be suggesting that If I want to use a role at all, it should be `dialog` and not `menu` for my use case. Is this understanding correct ?

Would you also say the same for the twitter use case as attached in my previous comments ?

**Comment by mcking65** [comment]
@gaurav5430 wrote:

> You are suggesting to use neither the `menu` role or the `dialog` role for my use case.

No. If you are making a popup,you need to use a dialog in your case.

> If that is the case, would using `aria-expanded` and focusing on the popup when opened be enough for a screenreader user?

If you are just showing/hiding content inline, and that content immediately follows the button in the DOM (and thus immediately follows in the reading order), then you could instead use the [disclosure pattern](https://w3c.github.io/aria-practices/#disclosure).

If you use disclosure, then do not move focus into the new content. Leave it on the button so the user can show then hide. This will also effectively communicate the state change to screen readers.

> You also seem to be suggesting that If I want to use a role at all, it should be `dialog` and not `menu` for my use case. Is this understanding correct ?

Yes.

> Would you also say the same for the twitter use case as attached in my previous comments ?

I can't comment on that because I don't have enough information about the scenario. I can't see the screen shots.

**Comment by gaurav5430** [comment]
Thanks for all your input, my content is in a popup, so I will use the dialog role.

Apologies, for the twitter screenshots, I had no easier way to actually describe what is going on in the code screenshot because that is what I am trying to understand as well, but let me give it a try.

Twitter has a popup which has 2 sections, a top section which just shows static content with an image, and a bottom section which is a list of links. They expose this popup with the `menu` role on the popup contianer and the `menuitem` role on each of the links and the haspopup='menu' on the trigger. They do not add the `menuitem` role to the static text, and I am assuming that would get removed from the accessibility tree, and to the screenreader user it will just be a menu with menuitems and the static text would not be exposed at all.


As for the disclosure pattern, I understand from your comments that it is best to use for content that is:
- inline and
- follows the disclosure trigger in the DOM
